### PR TITLE
[Retiarii] refactor based on the new launch approach

### PR DESCRIPTION
**Pull request by QuanluZhang**
- [x] make example code executable with the new launch approach
- [x] use decorator instead of `with` statement
- [x] small refactor of code gen on the new launch approach
- [x] refactor of strategy interface: merge trainer into base model
- [x] support darts example with the new launch approach (classic mode)
- [x] refactor the implementation of LayerChoice/InputChoice
- [x] LayerChoice/InputChoice backward compatibility
- [ ] refactor strategy to support trial concurrency
- [ ] unify launch style of classic mode and weight sharing for LayerChoice/InputChoice
- [ ] fix pylint
- [ ] support loop in graph ir (graph_gen)
- [ ] support remote training service

### CODE REVIEW

File: nni/experiment/pipe.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def __init__(self, experiment_id: str):
 
         def connect(self) -> BufferedIOBase:
             conn, _ = self._socket.accept()
-            self.file = conn.makefile('w+b')
+            self.file = conn.makefile('rwb')
```
Comment (ultmaster) [inline nni/experiment/pipe.py:55]: Why is this?

File: nni/retiarii/experiment.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    def __init__(self, training_service_platform: Optional[str] = None, **kwargs):
+        super().__init__(**kwargs)
+        if training_service_platform is not None:
+            assert 'training_service' not in kwargs
+            self.training_service = util.training_service_config_factory(training_service_platform)
+
+    def validate(self, initialized_tuner: bool = False) -> None:
+        super().validate()
+
+## End of public API ##
```
Comment (ultmaster) [inline nni/retiarii/experiment.py:46]: I don't think this is the end of public API?

File: nni/retiarii/experiment.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        Run the experiment.
+        This function will block until experiment finish or error.
+        """
+        self.config = config
+        self.start(config, port, debug)
+        try:
+            while True:
+                time.sleep(10)
+                status = self.get_status()
+                if status in ['ERROR', 'STOPPED', 'NO_MORE_TRIAL']:
```
Comment (ultmaster) [inline nni/retiarii/experiment.py:153]: What if the experiment is done?

File: nni/retiarii/nn/pytorch/nn.py [hunk 4]
```diff
@@ -4,30 +4,36 @@
...
-def get_records():
-    global _records
-    return _records
-
-def add_record(name, value):
-    global _records
-    if _records is not None:
-        assert name not in _records, '{} already in _records'.format(name)
-        _records[name] = value
+__all__ = [
```
Comment (ultmaster) [inline nni/retiarii/nn/pytorch/nn.py:12]: Is this  `nn.__all__` plus a few others?

File: nni/retiarii/nn/pytorch/nn.py [hunk 5]
```diff
@@ -108,24 +98,141 @@ def wrap_module(original_class):
...
 BatchNorm2d = wrap_module(nn.BatchNorm2d)
 ReLU = wrap_module(nn.ReLU)
 Dropout = wrap_module(nn.Dropout)
 Linear = wrap_module(nn.Linear)
 MaxPool2d = wrap_module(nn.MaxPool2d)
 AvgPool2d = wrap_module(nn.AvgPool2d)
 Identity = wrap_module(nn.Identity)
+AdaptiveAvgPool2d = wrap_module(nn.AdaptiveAvgPool2d)'''
+
+Identity = wrap_module(nn.Identity)
```
Comment (ultmaster) [inline nni/retiarii/nn/pytorch/nn.py:122]: Can we refactor it into a for loop?

File: nni/retiarii/converter/graph_gen.py [hunk 6]
```diff
@@ -463,6 +466,9 @@ def convert_module(script_module, module, module_name, ir_model):
 
     ir_graph._register()
 
+    if id(module) not in modules_arg:
+        raise RuntimeError(f'{original_type_name} arguments are not recorded, \
+            you may forget to decorate this class with @register_module()')
```
Comment (ultmaster) [inline nni/retiarii/converter/graph_gen.py:471]: you might have forgotten

File: nni/retiarii/nn/pytorch/nn.py (hunk 5 above)
Comment (QuanluZhang) [reply nni/retiarii/nn/pytorch/nn.py:122]: i think this way is more clear and easy to maintain

File: nni/retiarii/converter/graph_gen.py (hunk 6 above)
Comment (QuanluZhang) [reply nni/retiarii/converter/graph_gen.py:471]: fixed

File: nni/retiarii/nn/pytorch/nn.py (hunk 5 above)
Comment (QuanluZhang) [reply nni/retiarii/nn/pytorch/nn.py:122]: because not all the modules are included, for example, `loss` modules. and different versions of pytorch have different members in nn.module. I will handle different versions later

File: nni/retiarii/nn/pytorch/nn.py (hunk 4 above)
Comment (QuanluZhang) [reply nni/retiarii/nn/pytorch/nn.py:12]: yes, ours, for example placeholder

File: nni/experiment/pipe.py (hunk 1 above)
Comment (QuanluZhang) [reply nni/experiment/pipe.py:55]: https://github.com/microsoft/nni/pull/3183 @liuzhe-lz can answer this question

File: nni/retiarii/experiment.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/retiarii/experiment.py:47]: removed
